File: Problem_Solving/leetcode/linear_data_structure/Stack_and_Queue/316_Remove_Duplicate_Letters/removeDuplicateV2.cpp
#include <string>
#include <unordered_set>
#include <stack>

// fast
std::string removeDuplicateLetters(std::string s)
{
    int count[256] = {0};
    std::unordered_set<char> used;
    std::stack<char> result;

    for (char c : s) 
        count[c]++;

    for (char c : s)
    {
        count[c]--;

        if (used.count(c))
            continue;

        // 뒤에 붙일 문자가 있다면(중복문자) 스택에서 제거
        while (result.size() && c < result.top() && count[result.top()] > 0) 
        {
            used.erase(result.top());
            result.pop();
        }

        result.push(c);
        used.emplace(c);
    }

    std::string res;
    while (!result.empty())
    {
        res += result.top();
        result.pop();
    }
    std::reverse(res.begin(), res.end());

    return res;
}

int main() 
{
    std::string s("cbacdcbc");
    removeDuplicateLetters(s);
    return 0;
}

File: Problem_Solving/leetcode/linear_data_structure/Stack_and_Queue/316_Remove_Duplicate_Letters/removeDuplicate_v2.py
import collections

def removeDuplicateLetters(s: str) -> str:
    # 각 char의 숫자, 집합, 스택
    counter, seen, stack = collections.Counter(s), set(), []
    
    for char in s:
        counter[char] -= 1
        
        if char in seen:
            continue
        # 뒤에 붙일 문자가 남아 있다면 스택에서 제거(char가 2개 이상일때)
        while stack and char < stack[-1] and counter[stack[-1]] > 0:
            r = char < stack[-1]
            c1 = char
            c2 = stack[-1]
            seen.remove(stack.pop())
            
        stack.append(char)
        seen.add(char)
        
    return ''.join(stack)

s = "cbacdcbc"
removeDuplicateLetters(s)